[Workers] KV revision p2 (#3309)
cloudflare/cloudflare-docs: products/workers/src/content/runtime-apis/kv.md

You can also [write key-value pairs from the command line with Wrangler](/cli-wrangler/commands#kvkey) and [write data via the API](https://api.cloudflare.com/#workers-kv-namespace-write-key-value-pair).

> Due to the eventually consistent nature of Workers KV, concurrent writes from different edge locations can end up overwriting one another. It is a common pattern to write data via Wrangler or the API so that all writes initiate from the same location. This avoids competing, concurrent writes because of the single stream. All data is still readily available within all Workers bound to the namespace.

Reviewer: I think the emphasis on location is misleading here.  I think the original point was to avoid concurrent writes by using a single process that happens to use Wrangler or the API.  There's nothing inherent about Wrangler or the API that avoids concurrent writes..

> Due to the eventually consistent nature of Workers KV, concurrent writes from different edge locations can end up overwriting one another. It is a common pattern to write data via Wrangler or the API so that all writes initiate from the same location. This avoids competing, concurrent writes because of the single stream. All data is still readily available within all Workers bound to the namespace.

Reviewer 2: ```suggestion
Due to the eventually consistent nature of Workers KV, concurrent writes from different edge locations can end up overwriting one another. It is a common pattern to write data via Wrangler or the API so that all writes initiate from a single source. This avoids competing, concurrent writes because of the single stream. All data is still readily available within all Workers bound to the namespace.
```

Reviewer: ```suggestion
Due to the eventually consistent nature of Workers KV, concurrent writes can end up overwriting one another. It is a common pattern to write data from a single process via Wrangler or the API. This avoids competing, concurrent writes because of the single stream. All data is still readily available within all Workers bound to the namespace.
```

Writes are immediately visible to other requests in the same edge location, but can take up to 60 seconds to be visible in other parts of the world. Refer to [How KV works](/learning/how-kv-works) for more information on this topic.

#### Writing data in bulk

You can [write more than one key-value pair at a time with Wrangler](/cli-wrangler/commands#kvbulk) or [via the API](https://api.cloudflare.com/#workers-kv-namespace-write-multiple-key-value-pairs). The bulk API can accept up to 10,000 KV pairs at once. 

A `key` and `value` are required for each KV pair. The entire request size must be less than 100 megabytes. As of January 2022, Cloudflare does not support bulk writes from within a Worker.

#### Expiring keys

> Many common uses of Workers KV involve writing keys that are only meant to be valid for a certain amount of time. Rather than requiring applications to remember to delete such data at the appropriate time, Workers KV offers the ability to create keys that automatically expire. Expiration can occur either at a particular point in time or after a certain amount of time has passed since the key was last modified.

Reviewer: I'd say "Expiration can be specified to occur", as it is it's not clear whether I have control over it

> Many common uses of Workers KV involve writing keys that are only meant to be valid for a certain amount of time. Rather than requiring applications to remember to delete such data at the appropriate time, Workers KV offers the ability to create keys that automatically expire. Expiration can occur either at a particular point in time or after a certain amount of time has passed since the key was last modified.

Reviewer 2: ```suggestion
Many common uses of Workers KV involve writing keys that are only meant to be valid for a certain amount of time. Rather than requiring applications to remember to delete such data at the appropriate time, Workers KV offers the ability to create keys that automatically expire. You may configure expiration to occur either at a particular point in time or after a certain amount of time has passed since the key was last modified.
```

```

The `cacheTtl` parameter must be an integer that is greater than or equal to 60. It defines the length of time in seconds that a KV result is cached in the edge location that it is accessed from. This can be useful for reducing cold read latency on keys that are read relatively infrequently. It is especially useful if your data is write-once or write-rarely. It is not recommended if your data is updated often and you need to see updates shortly after they are written, because writes that happen from other edge locations will not be visible until the cached value expires.

The effective Cache TTL of an already cached item can be reduced by getting it again with a lower `cacheTtl`. For example, if you did `NAMESPACE.get(key, {cacheTtl: 86400})` but later realized that caching for 24 hours was too long, you could `NAMESPACE.get(key, {cacheTtl: 300})` or even `NAMESPACE.get(key)` and it would check for newer data to respect the provided `cacheTtl`, which defaults to `60` seconds.

#### Metadata

You can get the metadata associated with a key-value pair alongside its value by calling the `getWithMetadata` method on a namespace you have bound in your script:

```js

> const { value, metadata } = await NAMESPACE.getWithMetadata(key)

Reviewer: In the past we explicitly avoided destructuring assignment in docs to avoid confusing people who weren't familiar with it; are we walking that decision back?

Reviewer 2: Yes
